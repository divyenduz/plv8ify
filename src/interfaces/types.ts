const TYPES = {
  PLV8ify: Symbol('PLV8ify'),
  Bundler: Symbol('Bundler'),
  TSCompiler: Symbol('TSCompiler'),
}

export default TYPES
