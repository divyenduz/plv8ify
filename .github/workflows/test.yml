name: Run Tests and Release
# This workflow is triggered on pushes to the repository.
on:
  push:
    branches:
      - main
  pull_request:
    branches:
      - main
  workflow_dispatch:
    inputs:

env:
  NO_COLOR: true
  CI: 1

jobs:
  build:
    name: Test
    timeout-minutes: 10
    runs-on: ubuntu-latest
    steps:
      - name: Cancel Previous Runs
        uses: styfle/cancel-workflow-action@0.9.1
        with:
          access_token: ${{ github.token }}
      - uses: actions/checkout@v2
      - uses: oven-sh/setup-bun@v1

      - name: Turnstyle
        uses: softprops/turnstyle@v1
        env:
          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}

      - name: Install Dependencies
        run: echo "Installing dependencies" && bun install && echo "Done installing dependencies"

      - name: Run Types Check
        run: bun x tsc --noEmit

      - name: Run Tests
        run: bun test
