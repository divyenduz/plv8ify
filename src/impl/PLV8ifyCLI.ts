import fs from 'fs'
import { inject, injectable } from 'inversify'
import 'reflect-metadata'
import TYPES from '../interfaces/types'
import { match } from 'ts-pattern'

interface GetPLV8SQLFunctionArgs {
  fn: TSFunction
  scopePrefix: string
  pgFunctionDelimiter: string
  mode: Mode
  bundledJs: string
  fallbackReturnType: string
  defaultVolatility: Volatility
}

interface GetInitSQLFunctionArgs {
  fn: TSFunction
  scopePrefix: string
  bundledJs: string
  volatility: Volatility
}

// TODO: fixme, this is exported only for tests, is that needed?

@injectable()
export class PLV8ifyCLI implements PLV8ify {
  @inject(TYPES.Bundler) private _bundler: Bundler;
  @inject(TYPES.TSCompiler) private _tsCompiler: TSCompiler;

  private _typeMap = {
    number: 'float8',
    string: 'text',
    boolean: 'boolean',
  }

  init(inputFilePath: string) {
    this._tsCompiler.createSourceFile(inputFilePath)
  }

  async build({ mode, inputFile, scopePrefix }: BuildArgs) {
    const bundledJs = await this._bundler.bundle({
      inputFile,
      globalName: scopePrefix,
    })
    const modeAdjustedBundledJs = match(mode)
      .with('inline', () => bundledJs)
      .with('start_proc', () =>
        // Remove var from var plv8ify to make it attach to the global scope in start_proc mode
        bundledJs.replace(`var ${scopePrefix} =`, `this.${scopePrefix} =`)
      )
      .with('bundle', () =>
        // Remove var from var plv8ify to make it attach to the global scope in start_proc mode
        bundledJs.replace(`var ${scopePrefix} =`, `globalThis.${scopePrefix} =`)
      )
      .exhaustive()
    return modeAdjustedBundledJs
  }

  private writeFile(filePath: string, content: string) {
    try {
      fs.unlinkSync(filePath)
    } catch (e) {}
    fs.writeFileSync(filePath, content)
  }

  write(path: string, string: string) {
    this.writeFile(path, string)
  }

  private getScopedName(fn: TSFunction, scopePrefix: string) {
    const scopedName = scopePrefix + '_' + fn.name
    return scopedName
  }

  private getFileName(
    outputFolder: string,
    fn: TSFunction,
    scopePrefix: string
  ) {
    const scopedName = this.getScopedName(fn, scopePrefix)
    return `${outputFolder}/${scopedName}.plv8.sql`
  }

  private getFunctions() {
    return this._tsCompiler.getFunctions().map((fn) => {
      return {
        ...fn,
        returnType: this._typeMap[fn.returnType],
      }
    })
  }

  private getExportedFunctions() {
    return this.getFunctions().filter((fn) => fn.isExported)
  }

  private getFunctionVolatility(fn: TSFunction, defaultVolatility: Volatility) {
    const volatilityStr = '//@plv8ify-volatility-'
    const comments = fn.comments
    const volatility = (comments
      .filter((comment) => comment.includes(volatilityStr))
      .map((comment) => comment.replace(volatilityStr, ''))[0] ||
      defaultVolatility) as Volatility
    return volatility
  }

  // Input: parsed parameters, output of FunctionDeclaratioin.getParameters()
  // Output: SQL string of bind params
  private getSQLParametersString(
    parameters: TSFunctionParameter[],
    fallbackReturnType: string
  ) {
    return parameters
      .map((p) => {
        const { name, type } = p
        const mappedType = this._typeMap[type] || fallbackReturnType
        return `${name} ${mappedType}`
      })
      .join(',')
  }

  private getJSParametersString(parameters: TSFunctionParameter[]) {
    return parameters
      .map((p) => {
        const { name } = p
        return `${name}`
      })
      .join(',')
  }

  getPLV8SQLFunctions({
    scopePrefix,
    pgFunctionDelimiter,
    mode,
    bundledJs,
    fallbackReturnType,
    defaultVolatility,
    outputFolder,
  }: GetPLV8SQLFunctionsArgs) {
    const fns = this.getExportedFunctions()
    const sqls = fns.map((fn) => {
      return {
        filename: this.getFileName(outputFolder, fn, scopePrefix),
        sql: this.getPLV8SQLFunction({
          fn,
          scopePrefix,
          pgFunctionDelimiter,
          mode,
          bundledJs,
          fallbackReturnType,
          defaultVolatility,
        }),
      }
    })

    let startProcSQLs = []
    if (mode === 'start_proc' || mode === 'bundle') {
      // -- PLV8 + Server
      const initFunctionName = 'init'
      const virtualInitFn = {
        name: initFunctionName
      } as TSFunction // TODO: fixme, risky because it doesn't have all the properties of a virtual function

      const initFunction = this.getInitSQLFunction({
        fn: virtualInitFn,
        scopePrefix,
        bundledJs,
        volatility: defaultVolatility,
      })
      const initFileName = this.getFileName(
        outputFolder,
        virtualInitFn,
        scopePrefix
      )
      startProcSQLs.push({
        filename: initFileName,
        sql: initFunction,
      })
    }

    if (mode === 'start_proc') {
      const startFunctionName = 'start'
      const virtualStartFn = {
        name: startFunctionName
      } as TSFunction // TODO: fixme, risky because it doesn't have all the properties of a virtual function
      const startProcSQLScript = this.getStartProcSQLScript({scopePrefix})
      const startProcFileName = this.getFileName(
        outputFolder,
        virtualStartFn,
        scopePrefix
      )
      startProcSQLs.push({
        filename: startProcFileName,
        sql: startProcSQLScript,
      })
    }

    return sqls.concat(startProcSQLs)
  }

  getPLV8SQLFunction({
    fn,
    scopePrefix,
    pgFunctionDelimiter,
    mode,
    bundledJs,
    fallbackReturnType,
    defaultVolatility,
  }: GetPLV8SQLFunctionArgs) {
    const scopedName = scopePrefix + '_' + fn.name
    const sqlParametersString = this.getSQLParametersString(
      fn.parameters,
      fallbackReturnType
    )
    const jsParametersString = this.getJSParametersString(fn.parameters)
    const volatility = this.getFunctionVolatility(fn, defaultVolatility)
    const returnType = fn.returnType || fallbackReturnType

    return [
      `DROP FUNCTION IF EXISTS ${scopedName}(${sqlParametersString});`,
      `CREATE OR REPLACE FUNCTION ${scopedName}(${sqlParametersString}) RETURNS ${returnType} AS ${pgFunctionDelimiter}`,
      match(mode)
        .with('inline', () => bundledJs)
        .with('bundle', () => 
          `if (globalThis.${scopePrefix} === undefined) plv8.execute('SELECT ${scopePrefix}_init();');`
        )
        .otherwise(() => ''),
      `return ${scopePrefix}.${fn.name}(${jsParametersString})`,
      '',
      `${pgFunctionDelimiter} LANGUAGE plv8 ${volatility} STRICT;`,
    ].join('\n')
  }

  // TODO: fixme, can this be replaced with getPLV8SQLFunction
  private getInitSQLFunction({ fn, scopePrefix, bundledJs, volatility }: GetInitSQLFunctionArgs) {
    const scopedName = scopePrefix + '_' + fn.name
    return `DROP FUNCTION IF EXISTS ${scopedName}();
CREATE OR REPLACE FUNCTION ${scopedName}() RETURNS VOID AS $$
${bundledJs}
$$ LANGUAGE plv8 ${volatility} STRICT;
`
  }

  private getStartProcSQLScript = ({scopePrefix}) =>
  `
SET plv8.start_proc = ${scopePrefix}_init;
SELECT plv8_reset();
`
}
