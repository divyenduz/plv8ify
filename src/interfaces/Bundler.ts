interface BundleArgs {
  inputFile: string
  globalName: string // TODO: fixme, this is a part of esbuild config, ideally this shouldn't be a part of the common interface (unless all bundlers use this)
}

interface Bundler {
  bundle: (options: BundleArgs) => Promise<string>
}
