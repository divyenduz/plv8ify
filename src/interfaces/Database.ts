import { Context } from 'effect'

import { Database } from '../helpers/Database.js'

export interface DatabaseLayer {
  databaseUrl: string
  database: Database
}

export const DatabaseLayer = Context.Tag<DatabaseLayer>()
