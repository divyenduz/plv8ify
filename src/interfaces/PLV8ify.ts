type Volatility = 'VOLATILE' | 'STABLE' | 'IMMUTABLE'
type Mode = 'inline' | 'start_proc'

interface BuildArgs {
  mode: Mode
  inputFile: string
  scopePrefix: string
}

interface GetPLV8SQLFunctionsArgs {
  scopePrefix: string
  pgFunctionDelimiter: string
  mode: Mode
  bundledJs: string
  fallbackReturnType: string
  defaultVolatility: Volatility
  outputFolder: string
}

interface PLV8ify {
  init(inputFilePath: string): void
  build: (options: BuildArgs) => Promise<string>
  write: (path: string, string: string) => void

  getPLV8SQLFunctions({
    mode,
    scopePrefix,
    pgFunctionDelimiter,
    fallbackReturnType,
    bundledJs,
  }: GetPLV8SQLFunctionsArgs): {
    filename: string
    sql: string
  }[]
}
