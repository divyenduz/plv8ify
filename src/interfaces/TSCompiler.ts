export interface TSFunctionParameter {
  name: string
  type: string
}

export interface TSFunction {
  name: string
  isExported: boolean
  parameters: TSFunctionParameter[]
  returnType: string
  jsdocTags: { name: string, commentText: string }[]
}

export interface TSCompiler {
  createSourceFile(inputFilePath: string)

  getFunctions: () => TSFunction[]
}
