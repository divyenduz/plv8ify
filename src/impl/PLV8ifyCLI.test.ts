import { describe, expect, it } from 'bun:test'
import { TSFunction } from 'src/interfaces/TSCompiler'

import { PLV8ifyCLI } from './PLV8ifyCLI'

describe('PLV8ifyCLI tests', () => {
  it('getSQLFunction with parameters', async () => {
    const plv8ify = new PLV8ifyCLI()
    const sql = plv8ify.getPLV8SQLFunction({
      fn: {
        name: 'test',
        isExported: true,
        parameters: [],
        returnType: 'void',
        jsdocTags: []
      },
      scopePrefix: 'plv8ify_',
      mode: 'inline',
      defaultVolatility: 'IMMUTABLE',
      bundledJs: `console.log('hello')`,
      pgFunctionDelimiter: '$plv8ify$',
      fallbackReturnType: 'JSONB',
    })
    expect(sql).toMatchSnapshot()
  })

  it('getSQLFunction with custom delimiter', async () => {
    const plv8ify = new PLV8ifyCLI()
    const sql = plv8ify.getPLV8SQLFunction({
      fn: {
        name: 'test',
        isExported: true,
        parameters: [],
        returnType: 'void',
        jsdocTags: [],
      },
      scopePrefix: 'plv8ify_',
      mode: 'inline',
      defaultVolatility: 'IMMUTABLE',
      bundledJs: `console.log('hello')`,
      pgFunctionDelimiter: '$function$',
      fallbackReturnType: 'JSONB',
    })

    expect(sql).toMatchSnapshot()
  })

  it('getSQLFunction with parameters-trigger', async () => {
    const plv8ify = new PLV8ifyCLI()
    const sql = plv8ify.getPLV8SQLFunction({
      fn: {
        name: 'test',
        isExported: true,
        parameters: [
          { name: 'NEW', type: 'testRow' },
          { name: 'OLD', type: 'testRow' },
        ],
        returnType: 'object',
        jsdocTags: [
          { name: 'plv8ify_trigger', commentText: '' },
        ]
      },
      scopePrefix: 'plv8ify_',
      mode: 'inline',
      defaultVolatility: 'IMMUTABLE',
      bundledJs: `
function test(NEW, OLD) {
  if (TG_OP === "UPDATE") {
    NEW.event_name = NEW.event_name ?? OLD.event_name;
    return NEW;
  }
  if (TG_OP === "INSERT") {
    NEW.id = 102;
    return NEW;
  }
}      
      `,
      pgFunctionDelimiter: '$plv8ify$',
      fallbackReturnType: 'JSONB',
    })
    expect(sql).toMatchSnapshot()
  })

  it('getSQLFunction with custom-schema', async () => {
    const plv8ify = new PLV8ifyCLI()
    const sql = plv8ify.getPLV8SQLFunction({
      fn: {
        name: 'test',
        isExported: true,
        parameters: [],
        returnType: 'string',
        jsdocTags: [
          { name: 'plv8ify_schema_name', commentText: 'testschema' },
        ]
      },
      scopePrefix: 'plv8ify_',
      mode: 'inline',
      defaultVolatility: 'IMMUTABLE',
      bundledJs: `
function test() {
  return "hello";
}      
      `,
      pgFunctionDelimiter: '$plv8ify$',
      fallbackReturnType: 'JSONB',
    })
    expect(sql).toMatchSnapshot()
  })

  it('getSQLFunction with custom type', async () => {
    const plv8ify = new PLV8ifyCLI()
    plv8ify.init('', './src/test-fixtures/types-custom.fixture.js')

    const sql = plv8ify.getPLV8SQLFunction({
      fn: {
        name: 'test',
        isExported: true,
        parameters: [{ name: 'test', type: 'test_type[]' }],
        returnType: 'object',
        jsdocTags: []
      },
      scopePrefix: '',
      mode: 'inline',
      defaultVolatility: 'IMMUTABLE',
      bundledJs: `
      export function hello(test: test_type[]) {
        return {
          name: "Hello" + test[0].name,
          age: test[0].age,
        }
      }      
      `,
      pgFunctionDelimiter: '$plv8ify$',
      fallbackReturnType: 'JSONB',
    })
    expect(sql).toMatchSnapshot()
  })
})
