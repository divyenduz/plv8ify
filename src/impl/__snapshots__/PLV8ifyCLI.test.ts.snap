// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`PLV8ifyCLI tests getSQLFunction with parameters 1`] = `
"DROP FUNCTION IF EXISTS plv8ify_test();
CREATE OR REPLACE FUNCTION plv8ify_test() RETURNS JSONB AS $plv8ify$
console.log('hello')
return test()

$plv8ify$ LANGUAGE plv8 IMMUTABLE STRICT;"
`;

exports[`PLV8ifyCLI tests getSQLFunction with delimiter 1`] = `
"DROP FUNCTION IF EXISTS plv8ify_test();
CREATE OR REPLACE FUNCTION plv8ify_test() RETURNS JSONB AS $function$
console.log('hello')
return test()

$function$ LANGUAGE plv8 IMMUTABLE STRICT;"
`;


exports[`PLV8ifyCLI tests getSQLFunction with parameters-trigger 1`] = `
"DROP FUNCTION IF EXISTS plv8ify_test();
CREATE OR REPLACE FUNCTION plv8ify_test() RETURNS TRIGGER AS $plv8ify$

function test(NEW, OLD) {
  if (TG_OP === "UPDATE") {
    NEW.event_name = NEW.event_name ?? OLD.event_name;
    return NEW;
  }
  if (TG_OP === "INSERT") {
    NEW.id = 102;
    return NEW;
  }
}      
      
return test(NEW,OLD)

$plv8ify$ LANGUAGE plv8 IMMUTABLE STRICT;"
`;



exports[`PLV8ifyCLI tests getSQLFunction with custom type 1`] = `
"DROP FUNCTION IF EXISTS plv8ify_test(test test_type[]);
CREATE OR REPLACE FUNCTION plv8ify_test(test test_type[]) RETURNS JSONB AS $plv8ify$

      export function hello(test: test_type[]) {
        return {
          name: "Hello" + test[0].name,
          age: test[0].age,
        }
      }      
      
return test(test)

$plv8ify$ LANGUAGE plv8 IMMUTABLE STRICT;"
`;
